src/language-js/printer-estree.js

```diff
@@ -1726,7 +1726,7 @@ function printPathNoParens(path, options, print, args) {
       const hasValue = n.declarations.some(decl => decl.init);
 
       let firstVariable;
-      if (printed.length === 1 && !n.declarations[0].comments) {
+      if (printed.length === 1 && !hasLeadingComment(n.declarations[0])) {
```

Comment:
This condition is for code like below:
```js
const // comment
  foo = 3;
```
So, only target the leading comment.

Reply:
But why does it happen only with `typescript`?

Reply:
I found that the our comment attachment for babel's AST and ts-estree's AST is different in this case.

- For babel's AST: There is `n.comments`, there is not `n.declarations[0].comments`. (but, `n.declarations[0].trailingComments` does exists.)
- For ts-estree's AST: There is `n.declarations[0].comments`, there is not `n.comment`.

Probably this is a bug, I think we should fix comment attachment for Babel's AST.

Reply:
I investigated it a bit too. This difference is caused by the difference in location data for the `VariableDeclaration` node between the parsers.

**Prettier pr-7094**
[Playground link](https://deploy-preview-7094--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuc0DOMAEAzCFMC8mAFAJSEB8mwAOlJpgE5wwCujUA3HQL50D0-HHjrcoIADQgIABxgBLdMlABDRowgB3AApqEaZCBUA3CPIAmkkACNGKsAGsWAZRn35UAObIYjVnCkACxgAWwAbAHVA+Xg0NzA4Z30Y+WMYgE9DMDQDKQ80OEYYbTtPEJVkbBUwgqkAKzQADwAhO0cXFRC4ABkPOErq2pAGxucPTzC4AEVWCHgBmoCQN0YCxkMYdJk4NDBGeTkrGX3YCIsYQOQADgAGKWOIAoi7GUNjncLjfqkAR1n4EqyAwoFRoAC0UDgcHM0KszD+8mYJRUZQqSCqiykBRC8h8fiWaHGkxmc366MGSxgKmsZ3MF2QACYpL4VPIwuMAMIQELlQxQaDfECsAoAFWpwIxQ2M-gAklAYbBnHsDjAAILy5ybSYLAo8HhAA)
```sh
--parser typescript
```

**Input:**
```tsx
const foo = () => {
  return;
}
// foo
;

```

With `--parser typescript` it's 0...40. With `--parser babel` and `flow` it's 0...31.

Reply:
Turn's out there are already some hacks for this in `postprocess.js`. Must be very related.

Reply:
See `overrideLocEnd` there. Try replacing `options.parser === "flow"` with `Array.isArray(toBeOverriddenNode.range)` to make it work for TS too.

Reply:
Thanks, I see. done.

Reply:
So this change in `printer-estree.js` is not needed now.